framework/src/actions/AutoCheckpointAction.C

```diff
@@ -0,0 +1,38 @@
+#include "AutoCheckpointAction.h"
+#include "FEProblem.h"
+#include "Checkpoint.h"
+registerMooseAction("MooseApp", AutoCheckpointAction, "auto_checkpoint_action");
+
+InputParameters
+AutoCheckpointAction::validParams()
+{
+  InputParameters params = Action::validParams();
+
+  return params;
+}
+
+AutoCheckpointAction::AutoCheckpointAction(const InputParameters & params) : Action(params) {}
+
+void
+AutoCheckpointAction::act()
+{
+  if (_app.isUltimateMaster())
+  {
+    // if there's already a checkpoint object, we don't need to worry about creating a new
+    // checkpoint
+    if (!_app.getOutputWarehouse().getOutputs<Checkpoint>().empty())
+    {
+
+      // Take the first checkpoint object, since we only need to/should make one object the autosave
+      _app.getOutputWarehouse().getOutputs<Checkpoint>()[0]->set_autosave_flag(
+          AUTOSAVE_TYPE::MODIFIED_EXISTING);
+      return;
+    }
+
+    // If there isn't an existing one, init a new one
+    auto cp_params = _factory.getValidParams("Checkpoint");
+    cp_params.setParameters("is_autosave", AUTOSAVE_TYPE::SYSTEM_AUTOSAVE);
+    cp_params.setParameters("file_base", std::string("autosave"));
```

Comment:
I know we talked about this the other day with the templates messing up the automatic conversion from `char *` to `std::string`. The other way you can fix this function is to do it like this:
`cp_params.setParameters<std::string>("file_base", "autosave"));`

Reply:
Is it fine as it is now or does need to be changed?

Reply:
It's fine - just showing you how to use templates.